tool/thingtalk-dataset.js

```diff
@@ -0,0 +1,312 @@
+        if (this._initialization === null)
+            this._initialization = this._initialize();
+
+        this._initialization.then(() => this._output()).catch((e) => {
+            console.error(e);
+            this.emit('error', e);
+        });
+    }
+    
+    _output() {
+        if (this._i == this._dataset.examples.length) {  // done
+            this.push(null);
+            return;
+        } else if (this._i == -1) {
+            this.push({
+                type: 'meta',
+                data: {
+                    name: this._dataset.name,
+                    language: this._dataset.language,
+                    annotations: this._dataset.annotations,
+                    num_examples: this._dataset.examples.length
+                }
+            });
+        } else {
+            this.push({
+                type: 'example',
+                data: this._dataset.examples[this._i]
+            });
+        }
+        this._i++;
+    }
+    
+    async _initialize() {
+        this._dataset = await this._loadDataset();
+        this._i = -1;
+        
+        if (this._options.debug) {
+            console.log('Loaded ' + this._dataset.examples.length + ' templates');
+        }
+    }
+    
+    async _loadDataset() {
+        const code = await this._tpClient.getAllExamples();
+        const parsed = await Grammar.parse(code);
+        return parsed.datasets[0];
+    }
+}
+
+class ThingTalkDatasetWriter extends stream.Writable {
+    constructor(options) {
+        super({ objectMode: true });
+        
+        this.outputStream = options.outputStream;
+        this.options = options;
+        
+        this._meta = null;
+        this._examples = [];
+    }
+    
+    _write(buf, enc, next) {
+        if (buf.type == 'meta') {
+            if(this._meta)
+                throw 'Meta data has be set.';
+            this._meta = buf.data;
+        } else if(buf.type == 'example') {
+            this._examples.push(buf.data);
+        } else {
+            throw `Unsupported buffer type: ${buf.type}`;
+        }
+        process.nextTick(next);
+    }
+    
+    _final(callback) {
+        const output = new Library([], [
+            new Dataset(
+                this._meta.name,
+                this._meta.language,
+                this._examples,
+                this._meta.annotations
+            )
+        ]);
+        this.outputStream.write(output.prettyprint());
+        
+        this._meta = null;
+        this._examples = [];
+    }
+}
+
+module.exports = {
+    ThingTalkDatasetCleaner,
+    ThingTalkDatasetPreprocessor,
+    ThingTalkDatasetReader,
+    ThingTalkDatasetWriter
+};
+
+/* The following part is modified from Almond-Cloud
+ * https://github.com/stanford-oval/almond-cloud/blob/master/util/tokenize.js
+ * https://github.com/stanford-oval/almond-cloud/blob/master/util/validation.js
+ */
+const PARAM_REGEX = /\$(?:\$|([a-zA-Z0-9_]+(?![a-zA-Z0-9_]))|{([a-zA-Z0-9_]+)(?::([a-zA-Z0-9_-]+))?})/;
```

Comment:
It would be good to have all the copied over code into a library, so that almond-cloud can switch to the copy of the code in Genie.
Also, if you copy code you must either add a copyright notice, or just add shared copyright at the top.

Reply:
Okay I'll put them in a library and write the correct copyright. Thanks.